File: titan-refactor.mjs
import * as fs from 'fs';
import * as path from 'path';
import * as esprima from 'esprima';

const targetDir = './basic-logic/algorithms';

function getExports(code) {
    const exportsList = [];
    try {
        const ast = esprima.parseModule(code);

        ast.body.forEach(node => {
            if (node.type === 'ExportNamedDeclaration' && node.declaration) {
                // export const foo = (args) => {}
                if (node.declaration.type === 'VariableDeclaration') {
                    node.declaration.declarations.forEach(decl => {
                        if (decl.id.type === 'Identifier' && decl.init) {
                            const funcName = decl.id.name;
                            let params = [];

                            // Arrow Function or standard Function
                            if (decl.init.type === 'ArrowFunctionExpression' || decl.init.type === 'FunctionExpression') {
                                params = decl.init.params.map(p => {
                                    if (p.type === 'Identifier') return p.name;
                                    if (p.type === 'AssignmentPattern' && p.left.type === 'Identifier') return p.left.name;
                                    if (p.type === 'RestElement' && p.argument.type === 'Identifier') return '...' + p.argument.name;
                                    return 'arg'; // Fallback
                                });
                            }
                            exportsList.push({ name: funcName, args: params });
                        }
                    });
                }
                // export function foo(args) {}
                else if (node.declaration.type === 'FunctionDeclaration') {
                    const funcName = node.declaration.id.name;
                    const params = node.declaration.params.map(p => {
                        if (p.type === 'Identifier') return p.name;
                        if (p.type === 'AssignmentPattern' && p.left.type === 'Identifier') return p.left.name;
                        if (p.type === 'RestElement' && p.argument.type === 'Identifier') return '...' + p.argument.name;
                        return 'arg';
                    });
                    exportsList.push({ name: funcName, args: params });
                }
            }
        });
    } catch (e) {
        console.error("Esprima parsing error on code:", e.message);
    }
    return exportsList;
}

function refactorTitanLogic() {
    const manifest = [];
    const files = fs.readdirSync(targetDir);

    files.forEach(filename => {
        if (!filename.endsWith('.js') || filename === 'algorithms-manifest.json') return;

        const filepath = path.join(targetDir, filename);
        let content = fs.readFileSync(filepath, 'utf8');

        // Regex refactoring for standard function missing exports (porting functionality from old python script)
        const stdFuncPattern = /(?<!export\s)function\s+(\w+)\s*\((.*?)\)\s*\{/g;
        const arrowFuncPattern = /(?<!export\s)const\s+(\w+)\s*=\s*\((.*?)\)\s*=>\s*\{/g;

        let hasNewExports = false;

        if (!content.includes('export ')) {
            const newContent = content
                .replace(stdFuncPattern, 'export const $1 = function($2) {')
                // Using standard functions to retain `arguments` compatibility by default, rather than erroring arrow funcs
                .replace(arrowFuncPattern, 'export const $1 = ($2) => {');

            if (newContent !== content) {
                fs.writeFileSync(filepath, newContent, 'utf8');
                content = newContent;
                console.log(`[+] Refactored logic to export module: ${filename}`);
                hasNewExports = true;
            } else {
                console.log(`[!] No changes needed for target: ${filename}`);
            }
        }

        // Parse with Esprima to build manifest mappings
        const exportsMetadata = getExports(content);

        if (exportsMetadata.length > 0) {
            const title = filename.replace('.js', '').split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
            const type = exportsMetadata.length > 1 ? 'Interactive' : 'Simple';
            let description = '';

            if (type === 'Interactive') {
                description = `[Stateful Neural Pathway] Requires interactive parameter injection. Executes advanced multi-stage logic for ${title}.`;
            } else {
                description = `[Standard Heuristic Algorithm] Executes the purely functional protocol for ${title}. Supply the required parameters to compute the sequence.`;
            }

            manifest.push({
                filename: filename,
                type: type,
                exports: exportsMetadata,
                title: title,
                description: description
            });
        } else {
            console.warn(`[WARNING] No exports detected in ${filename} after parsing.`);
        }
    });

    // Write enhanced manifest
    const manifestPath = path.join(targetDir, 'algorithms-manifest.json');
    fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 4), 'utf8');
    console.log("[+] Enhanced Algorithm Manifest deployed with telemetry mapping.");
}

console.log("Scanning Neural Link for logic architectures...");
refactorTitanLogic();
console.log("All systems are primed and mapped.");

File: titan-refactor.py
import os
import re
import json

target_dir = './basic-logic/algorithms'

# Pattern 1: Standard 'function name(params) {'
std_func_pattern = re.compile(r'(?<!export\s)function\s+(\w+)\s*\((.*?)\)\s*\{')

# Pattern 2: Arrow 'const name = (params) => {'
arrow_func_pattern = re.compile(r'(?<!export\s)const\s+(\w+)\s*=\s*\((.*?)\)\s*=>\s*\{')

def refactor_titan_logic():
    manifest = []
    for filename in os.listdir(target_dir):
        if filename.endswith('.js'):
            filepath = os.path.join(target_dir, filename)
            
            # Add to manifest
            manifest.append(filename)

            with open(filepath, 'r') as f:
                content = f.read()

            # Skip if file already exports anything
            if 'export ' in content:
                print(f"[-] Skipping {filename}: Export detected.")
                continue

            # Check for Standard Functions -> convert to Export Const Arrow
            new_content = std_func_pattern.sub(r'export const \1 = (\2) => {', content)
            
            # Check for existing Arrow Functions -> add Export prefix
            new_content = arrow_func_pattern.sub(r'export const \1 = (\2) => {', new_content)

            if new_content != content:
                with open(filepath, 'w') as f:
                    f.write(new_content)
                print(f"[+] Refactored {filename}")
            else:
                print(f"[!] No changes needed for {filename}")

    # Write manifest
    manifest_path = os.path.join(target_dir, 'algorithms-manifest.json')
    with open(manifest_path, 'w') as f:
        json.dump(manifest, f, indent=4)
    print("[+] Algorithm Manifest deployed successfully.")

if __name__ == "__main__":
    print("Scanning Neural Link for legacy and modern logic...")
    refactor_titan_logic()
    print("All algorithms are now ready for tactical deployment.")